api/classes/class-lightweight-api.php

```diff
@@ -0,0 +1,303 @@
+	 *
+	 * @var suppress_all_newsletter_campaigns_if_one_dismissed
+	 */
+	public $suppress_all_newsletter_campaigns_if_one_dismissed;
+
+	/**
+	 * Campaign has Newsletter prompt.
+	 *
+	 * @var has_newsletter_prompt
+	 */
+	public $has_newsletter_prompt;
+
+	/**
+	 * Response object.
+	 *
+	 * @var response
+	 */
+	public $response = [];
+
+	/**
+	 * The referer URL.
+	 *
+	 * @var referer_url
+	 */
+	public $referer_url;
+
+	/**
+	 * Debugging information.
+	 *
+	 * @var debug
+	 */
+	public $debug;
+
+	/**
+	 * Constructor.
+	 */
+	public function __construct() {
+		$this->debug = [
+			'queries' => [],
+		];
+		if ( ! $this->verify_referer() ) {
+			$this->error( 'invalid_referer' );
+		}
+		$this->db              = $this->connect();
+		$this->rid             = ! empty( $_REQUEST['rid'] ) ? // phpcs:ignore
+			filter_input( INPUT_POST | INPUT_GET, 'rid', FILTER_SANITIZE_SPECIAL_CHARS ) :
+			null; // phpcs:ignore
+		$this->popup_id        = ! empty( $_REQUEST['popup_id'] ) ? // phpcs:ignore
+			filter_input( INPUT_POST | INPUT_GET, 'popup_id', FILTER_SANITIZE_SPECIAL_CHARS ) :
+			null; // phpcs:ignore
+		$this->frequency       = ! empty( $_REQUEST['frequency'] ) ? // phpcs:ignore
+			filter_input( INPUT_POST | INPUT_GET, 'frequency', FILTER_SANITIZE_SPECIAL_CHARS ) :
+			null; // phpcs:ignore
+		$this->placement       = ! empty( $_REQUEST['placement'] ) ? // phpcs:ignore
+			filter_input( INPUT_POST | INPUT_GET, 'placement', FILTER_SANITIZE_SPECIAL_CHARS ) :
+			null; // phpcs:ignore
+		$this->utm_suppression = ! empty( $_REQUEST['utm_suppression'] ) ? // phpcs:ignore
+			filter_input( INPUT_POST | INPUT_GET, 'utm_suppression', FILTER_SANITIZE_SPECIAL_CHARS ) :
+			null; // phpcs:ignore
+
+		$this->suppress_newsletter_campaigns                      = ! empty( $_REQUEST['suppress_newsletter_campaigns'] ) ? // phpcs:ignore
+			filter_input( INPUT_POST | INPUT_GET, 'suppress_newsletter_campaigns', FILTER_SANITIZE_SPECIAL_CHARS ) :
+			null; // phpcs:ignore
+		$this->suppress_all_newsletter_campaigns_if_one_dismissed = ! empty( $_REQUEST['suppress_all_newsletter_campaigns_if_one_dismissed'] ) ? // phpcs:ignore
+			filter_input( INPUT_POST | INPUT_GET, 'suppress_all_newsletter_campaigns_if_one_dismissed', FILTER_SANITIZE_SPECIAL_CHARS ) :
+			null; // phpcs:ignore
+		$this->has_newsletter_prompt = ! empty( $_REQUEST['has_newsletter_prompt'] ) ? // phpcs:ignore
+			filter_input( INPUT_POST | INPUT_GET, 'has_newsletter_prompt', FILTER_SANITIZE_SPECIAL_CHARS ) :
+			null; // phpcs:ignore
+
+		$this->referer_url = filter_input( INPUT_SERVER, 'HTTP_REFERER', FILTER_SANITIZE_STRING );
+
+		if ( 'inline' !== $this->placement && 'always' === $this->frequency ) {
+			$this->frequency = 'once';
+		}
+	}
+
+	/**
+	 * Verify referer is valid.
+	 */
+	public function verify_referer() {
+		$http_referer = ! empty( $_SERVER['HTTP_REFERER'] ) ? parse_url( $_SERVER['HTTP_REFERER'] , PHP_URL_HOST ) : null; // phpcs:ignore
+		$valid_referers = [
+			$http_referer,
```

Comment:
> or take them from config?

What do you mean?

Reply:
Sorry if I was unclear. What I meant was since the value of `$_SERVER['HTTP_REFERER']` [can be set by the client](https://thisinterestsme.com/php-spoof-http-referer-field/) as they wish, instead of just trusting the value of `$_SERVER['HTTP_REFERER']`, should we have a hard-coded list of valid referers (somewhere in config, or option, or else)?

Reply:
I wouldn't want to put it in the Database because this would add a query. If we put it in a config, this would suggest we need to go back to the original approach where the plugin has it's own custom config file.

Reply:
I understand, and also agree. I think that my suggestion to store valid HTTP_REFERERs maybe wasn't good at all. 

Thinking once again about it -- this is simple a form of authentication, right? We could probably elaborate and use something else.

Pinging @adekbadek -- Adam, since you have experience with implementing WPCOM authentication, do you think it would be feasible to implement the same kind of oAuth verification here? Briefly put -- this is a self-standing light-weight API with a couple of GET and POST endpoints, meant to run on Newspack Atomic sites. We're also looking for something fast, and easy on the resources.

Reply:
In other works, the proposition is to use a self-issued authentication token to verify that the API caller is in fact the site, right? That still would mean an additional query, as the token would have to be stored somewhere.

Reply:
https://github.com/Automattic/newspack-popups/pull/210 brings back the idea of a custom config file. If we go this way we could hardcode the hostname in the config, as @iuravic suggests.

Reply:
> In other words, the proposition is to use a self-issued authentication token to verify that the API caller is in fact the site, right? That still would mean an additional query, as the token would have to be stored somewhere. But other than that, it's feasible.

Won't someone just be able to copy the token by examining the API request?

Assuming all the other security bits are done well, I think the main "danger" with the API would be an attacker making hundreds of thousands of requests to fill up the database. I assume the system-level checks that prevent DoS attacks would catch that sort of attack, though.